compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -75,186 +75,507 @@ std::optional<uint32_t> getNumberIterations(scf::ForOp loop) {
-                                                        shimTileToGeneratorMap);
-        if (succeeded(tile) && *tile == tileOp) bdIdCount++;
+/// A DmaBatch contains the following :-
+///   1. DmaOps within the current batch.
+///   2. Required Bd Ids by the current batch.
+///   3. Nested DmaBatch chain.
+///   4. Pointer to the sibling DmaBatch.
+///   4. A pointer to the parent op if the parent of is a scf.for.
+struct DmaBatch {
+  SmallVector<AMDAIE::NpuDmaCpyNdOp> currentDmaOps = {};
+  int32_t requiredBdIds = 0;
+  SmallVector<std::shared_ptr<DmaBatch>> immediateInnerBatches = {};
+  std::shared_ptr<DmaBatch> nextDmaBatch = nullptr;
+  scf::ForOp forOpParent = nullptr;
+
+  /// A DmaBatch with no DmaOps and no nested DmaBatches is an empty DmaBatch.
+  bool isEmpty() const {
+    return (currentDmaOps.empty() && immediateInnerBatches.empty());
+  }
+
+  /// Given a DmaBatch `outerDmaBatch`, traverse each nested DmaBatch in it and
+  /// infer the required Bd Ids for them. If the nested DmaBatch is surrounded
+  /// with a scf.for, we account for that while sending the required Bd Ids to
+  /// the caller of this API.
+  int32_t getRequiredBdIdsForInnerBatches() const {
+    if (immediateInnerBatches.empty()) return 0;
+    int32_t requiredBdIds = 0;
+    for (const std::shared_ptr<DmaBatch> &dmaBatch : immediateInnerBatches) {
+      std::shared_ptr<DmaBatch> currDmaBatch = dmaBatch;
+      do {
+        if (currDmaBatch->isEmpty()) break;
+        int32_t totalDmaOpsInCurrentBatch = currDmaBatch->currentDmaOps.size();
+        int32_t requiredBdIdsForInnerBatches =
+            currDmaBatch->getRequiredBdIdsForInnerBatches();
+        currDmaBatch->requiredBdIds =
+            requiredBdIdsForInnerBatches + totalDmaOpsInCurrentBatch;
+        requiredBdIds += totalDmaOpsInCurrentBatch;
+        if (currDmaBatch->forOpParent) {
+          std::optional<uint32_t> numIterations =
+              getNumberIterations(currDmaBatch->forOpParent);
+          if (numIterations) {
+            requiredBdIds +=
+                requiredBdIdsForInnerBatches * numIterations.value();
+          } else {
+            requiredBdIds += requiredBdIdsForInnerBatches;
+          }
+        }
+        currDmaBatch = currDmaBatch->nextDmaBatch;
+      } while (currDmaBatch != nullptr);
+    }
+    return requiredBdIds;
+  }
+};
+
+/// `TileDmaBatchGraph` is a struct that maintains a mapping from a tile to its
+/// corresponding `DmaBatch`. It contains utilities that help form/update the
+/// `DmaBatch` graph for a particular tile.
+class TileDmaBatchGraph {
+  /// The main tile->DmaBatch graph.
+  llvm::MapVector<AMDAIE::TileOp, std::shared_ptr<DmaBatch>> tileDmaBatchGraph;
+  /// Helps keep track of the current DmaBatch of a tile that's being processed.
+  /// It holds the reference to the last DmaBatch for every tile on the main
+  /// graph.
+  DenseMap<AMDAIE::TileOp, std::shared_ptr<DmaBatch>> currentTileBatch;
```

Comment:
Do we really need both? I fail to see where `currentTileBatch` is used in the processing in `BdIdAssignmentUtil`?

Reply:
`currentTileBatch` was used to make the creation of tileDmaBatchGraph fairly simple as it involved node addition to a linkedlist, thus maintaining the pointer to the end.

But, I have now updated the algo to not use LinkedList paradigm anymore. Can you check now?